src/utils/src/Time.c

```diff
@@ -3,6 +3,70 @@
 getTime globalGetTime = defaultGetTime;
 getTime globalGetRealTime = defaultGetTime;
 
+STATUS generateTimestampStrInMilliseconds(PCHAR formatStr, PCHAR pDestBuffer, UINT32 destBufferLen, PUINT32 pFormattedStrLen)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 formattedStrLen;
+    UINT32 millisecondVal;
+    struct tm* timeinfo;
+
+#if defined _WIN32 || defined _WIN64
+    FILETIME fileTime;
+    GetSystemTimeAsFileTime(&fileTime);
+
+    // Convert the file time to 100-nanosecond intervals (since January 1, 1601)
+    ULARGE_INTEGER uli;
+    uli.LowPart = fileTime.dwLowDateTime;
+    uli.HighPart = fileTime.dwHighDateTime;
+    ULONGLONG fileTime100ns = uli.QuadPart;
+    fileTime100ns -= TIME_DIFF_UNIX_WINDOWS_TIME;
+
+    // Convert to milliseconds
+    ULONGLONG milliseconds100ns = fileTime100ns / 10000;
+
+    // Convert milliseconds to time_t
+    time_t seconds = milliseconds100ns / 1000;
+
+    // Convert time_t to UTC tm struct
+    timeinfo = GMTIME(&seconds);
+
+    millisecondVal = milliseconds100ns % 1000;
+
+#elif defined __MACH__ || defined __CYGWIN__
+    struct timeval tv;
+    gettimeofday(&tv, NULL);  // Get current time
+
+    timeinfo = GMTIME(&(tv.tv_sec));  // Convert to broken-down time
+
+    millisecondVal = tv.tv_usec / HUNDREDS_OF_NANOS_IN_A_MICROSECOND;  // Convert microseconds to milliseconds
+
+#else
+    struct timespec nowTime;
+    clock_gettime(CLOCK_REALTIME, &nowTime);
+
+    timeinfo = GMTIME(&(nowTime.tv_sec));
+    millisecondVal = nowTime.tv_nsec / DEFAULT_TIME_UNIT_IN_NANOS;  // Convert nanoseconds to milliseconds
+
+#endif
+    CHK(pDestBuffer != NULL, STATUS_NULL_ARG);
+    CHK(STRNLEN(formatStr, MAX_TIMESTAMP_FORMAT_STR_LEN + 1) <= MAX_TIMESTAMP_FORMAT_STR_LEN, STATUS_MAX_TIMESTAMP_FORMAT_STR_LEN_EXCEEDED);
+
+    formattedStrLen = 0;
+    *pFormattedStrLen = 0;
```

Comment:
Can we add a null check / bail out if `pFormattedStrLen` is NULL?  Especially since this is public API we cannot depend on clients to sanitize input.

I realize it's missing from `generateTimestampStr` function as well.

Reply:
Added.